### PR TITLE
`Development`: Use custom built images for GitLab and Jenkins

### PR DESCRIPTION
**Pull request by hhontheim**
#### General
- [ ] This is a small issue that I tested locally and was confirmed by another developer ~on a test server~. → Only local testing is possible here due to the creation of Docker containers for GitLab and Jenkins.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

### Motivation and Context
Due to the increasing number of developers in Artemis' developer team using a Mac with an M1 SoC, all images used for creating Docker containers must support `arm64` architectures. Unfortunately, GitLab does not provide one.

As a logical consequence, we should use custom-built images for Artemis that support those architectures.

### Description
This PR introduces the usage of custom-built images for GitLab and Jenkins from <https://github.com/ls1intum/Artemis-Local-Setup-Docker> (once ls1intum/Artemis-Local-Setup-Docker#4 has been merged).

Not only do they now provide native `arm64` versions – they additionally contain all changes such as installing plugins, already. This means, the images can directly be used in `docker-compose` files, such as [`gitlab-jenkins-mysql.yml`-file](https://github.com/ls1intum/Artemis/blob/feature/gitlab-jenkins-arm/src/main/docker/gitlab-jenkins-mysql.yml), eliminating the need to patch and build the images locally.

Additionally, the tag for the MySQL image is also changed to `8.0.28-oracle`. By changing it from images based on `debian:buster-slim` to those that base on `oraclelinux:8-slim`, we now use images that support both `amd64` as well as `arm64` architectures.

### Steps for Testing
1. [Follow the steps for setting up a local development environment with GitLab and Jenkins in the setup guide](https://artemis-platform.readthedocs.io/en/latest/dev/setup/jenkins-gitlab/).

> Please note that if the PR ls1intum/Artemis-Local-Setup-Docker#4 is not merged yet, you will not have access to the new images. In this case, either fork the repository [Artemis-Local-Setup-Docker](https://github.com/ls1intum/Artemis-Local-Setup-Docker) and merge the branch `feature/gitlab-jenkins-images` into `main` to build your own images, or use the ones for [GitLab](https://github.com/users/hhontheim/packages/container/package/artemis-gitlab) and [Jenkins](https://github.com/users/hhontheim/packages/container/package/artemis-jenkins) from my fork. Please use the `latest` tag – others might not be compatible...
>
> :warning: You will have to change the image names in the e.g. [`gitlab-jenkins-mysql.yml`-file](https://github.com/ls1intum/Artemis/blob/feature/gitlab-jenkins-arm/src/main/docker/gitlab-jenkins-mysql.yml) for Artemis to reflect the correct repository: `ghcr.io/`~`ls1intum`~`/artemis-{gitlab|jenkins}:latest` → `ghcr.io/`**`hhontheim`**`/artemis-{gitlab|jenkins}:latest` or your GH username.

2. Verify that your local setup is working. This includes (but is not limited to):

- User management with GitLab and Jenkins
- Programming exercises are usable

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

File: src/main/docker/jenkins.swift.override.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+version: "2"
+services:
+    bamboo:
+        image: ghcr.io/ls1intum/artemis-jenkins-swift:latest
```
Comment (sleiss) [inline src/main/docker/jenkins.swift.override.yml:4]: I am a bit confused by this line: Why do you specify a `jenkins` image for the `bamboo` service?

File: src/main/docker/gitlab-jenkins.yml [hunk 2]
```diff
@@ -25,15 +25,15 @@ services:
...
+                ipv4_address: 172.20.0.5
 
 networks:
     artemis:
         driver: "bridge"
         ipam:
             driver: default
             config:
-                - subnet: 172.19.0.0/24
+                - subnet: 172.20.0.0/24
```
Comment (sleiss) [inline src/main/docker/gitlab-jenkins.yml:36]: Was this change intended?
Now the subnets of the networks defined inside `atlassian.yml` and `gitlab-jenkins.yml` overlap, wheres `atlassian.yml` and `gitlab-jenkins-mysql.yml` do not overlap.

I think we should be consistent here (and maybe also remove `gitlab-jenkins.yml` or restructure it to remove duplication?).

File: src/main/docker/jenkins.swift.override.yml (hunk 1 above)
Comment (hhontheim) [reply src/main/docker/jenkins.swift.override.yml:4]: Good catch – this was just a copy-paste error. Will fix this!

**Comment by hhontheim** [comment]
While fixing @sleiss' comment, I wondered whether we actually need Jenkins Images with Swift pre-installed at all?
Especially since Swift-PEs will run (as all other exercises) on agents with their distinct docker containers...
The same goes for Bamboo's Swift image as well as for the `C`-variants...

What do you think about this, @Mtze?

File: src/main/docker/gitlab-jenkins.yml (hunk 2 above)
Comment (hhontheim) [reply src/main/docker/gitlab-jenkins.yml:36]: These changes actually came from @Hialus' contributions to this PR. But you're right that we should be consistent.

TBH, I'm not sure what's the best solution. On the one hand, having the same subnet for all configs might result in problems, if e.g. someone runs multiple configs (GitLab & Atlassian stack) at the same time (or at least creates the containers with compose). We would have to make sure, that the IP addresses don't collide. The IP `172.xxx.0.2`, e.g., is assigned to multiple services across all configs.

I would suggest that we stick to individual subnets for each service group. This way, they won't collide with each other at all.

Comment (hhontheim) [reply src/main/docker/gitlab-jenkins.yml:36]: Regarding the `gitlab-jenkins.yml` file:
I'm not quite sure what this file is intended for as I could not find any references to it in the documentation. But from its structure, it looks as if this file is for rather production-like systems, doesn't it?

If this is the case, we could aim for a "base" file that can be used in production (with CSRF enabled and no pre-set `master.key`) and provide an override file (like we do for the C and Swift images) that then can be used when running a cluster for local development.

But if the file is there for no valid reason, I think we should remove it. If someone doesn't need a DB, e.g., this service could just be commented out...

File: src/main/docker/gitlab-jenkins-mysql.yml [hunk 3]
```diff
@@ -41,9 +41,9 @@ services:
         environment:
             - MYSQL_ALLOW_EMPTY_PASSWORD=yes
             - MYSQL_DATABASE=Artemis
-        image: mysql:8.0.27
+        image: mysql:8.0.28-oracle
```
Comment (Mtze) [inline src/main/docker/gitlab-jenkins-mysql.yml:44]: Why do we have a MySQL container in this configuration anyway? To my understanding neither gitlab nor Jenkins use MySQL right?

File: docker-compose.yml [hunk 4]
```diff
@@ -40,7 +40,7 @@ services:
       - MYSQL_ALLOW_EMPTY_PASSWORD=yes
       - MYSQL_ROOT_PASSWORD=
       - MYSQL_DATABASE=Artemis
-    image: mysql:8.0.27
+    image: mysql:8.0.28-oracle
```
Comment (dfuchss) [inline docker-compose.yml:43]:
```suggestion
    image: mysql:8.0.29
```

File: docs/dev/setup.rst [hunk 5]
```diff
@@ -273,7 +273,7 @@ You could use a compose file similar to this (as an example this deployment uses
 
     services:
       gitlab:
-        image: gitlab/gitlab-ce
+        image: ghcr.io/ls1intum/artemis-gitlab:latest
```
Comment (dfuchss) [inline docs/dev/setup.rst:276]: I'm not sure whether the setup should refer to custom images of gitlab / jenkins
This would lead to maintenance effort for all depending images.

Maybe simply refer to a process on how to create the images on Mac

File: docs/dev/setup.rst [hunk 6]
```diff
@@ -304,7 +304,7 @@ You could use a compose file similar to this (as an example this deployment uses
           - artemis-net
 
       jenkins:
-        image: jenkins/jenkins:lts
+        image: ghcr.io/ls1intum/artemis-jenkins:latest
```
Comment (dfuchss) [inline docs/dev/setup.rst:307]: See above

File: docs/dev/setup.rst [hunk 7]
```diff
@@ -337,7 +337,7 @@ You could use a compose file similar to this (as an example this deployment uses
           - artemis-net
 
       artemis-db:
-        image: mysql:8.0.23
+        image: mysql:8.0.28-oracle
```
Comment (dfuchss) [inline docs/dev/setup.rst:340]:
```suggestion
        image: mysql:8.0.29
```

File: src/main/docker/gitlab-jenkins-mysql.yml (hunk 3 above)
Comment (dfuchss) [inline src/main/docker/gitlab-jenkins-mysql.yml:44]:
```suggestion
        image: mysql:8.0.29
```

File: src/main/kubernetes/artemis/statefulsets/artemis-mysql.yml [hunk 8]
```diff
@@ -31,7 +31,7 @@ spec:
         envFrom:
         - configMapRef:
             name: artemis-mysql
-        image: mysql:8.0.27
+        image: mysql:8.0.28-oracle
```
Comment (dfuchss) [inline src/main/kubernetes/artemis/statefulsets/artemis-mysql.yml:34]:
```suggestion
        image: mysql:8.0.29
```

File: src/main/docker/mysql.yml [hunk 9]
```diff
@@ -1,10 +1,7 @@
 version: '2.4'
 services:
     artemis-mysql:
-        # Use this for arm64 processors (and remove the platform: linux/x86_64
-        # image: ubuntu/mysql:8.0-21.10_beta
-        image: mysql:8.0.27
-        platform: linux/x86_64
+        image: mysql:8.0.28-oracle
```
Comment (dfuchss) [inline src/main/docker/mysql.yml:4]:
```suggestion
        image: mysql:8.0.29
```

**Review by dfuchss — CHANGES_REQUESTED** [review]
Some general questions. Also some version updates :)

**Comment by krusche** [comment]
We decided to close this PR with @Mtze and the deployment working group. The required implementation will be done in another PR.

**Comment by krusche** [comment]
Feel free to discuss this in the channel #artemis-deployment-wg in our Slack instance
